### ISSUE TITLE
Vim9: blocks don't work at the script level

### ISSUE THREAD

**Issue by lacygoill** [vim/vim#6498]
**Describe the bug**

In Vim9 script, blocks don't work at the script level.

**To Reproduce**

Run this shell command:

    vim -Nu NONE -S <(cat <<'EOF'
        vim9script
        {
            echo 'inside'
        }
        echo 'outside'
    EOF
    )

`E121` and `E492` are raised:

    Error detected while processing ...
    line    3:
    E121: Undefined variable: echo
    line    4:
    E492: Not an editor command:     }
    outside

**Expected behavior**

No error is raised.

**Environment**

 - Vim version: 8.2 Included patches: 1-1253
 - OS: Ubuntu 16.04.6 LTS
 - Terminal: XTerm(356)

**Additional context**

The issue disappears inside a `:def` function:
```vim
vim9script
def Func()
    {
        echo 'inside'
    }
    echo 'outside'
enddef
defcompile
```

**Comment by brammool** [comment]
> **Describe the bug**
> 
> Blocks don't work at the script level.
> 
> **To Reproduce**
> 
> Run this shell command:
> 
>     $ vim -Nu NONE -S <(cat <<'EOF'
>         vim9script
>         {
>             echo 'inside'
>         }
>         echo 'outside'
>     EOF
>     )
> 
> `E121` and `E492` are raised:
> 
>     Error detected while processing ...
>     line    3:
>     E121: Undefined variable: echo
>     line    4:
>     E492: Not an editor command:     }
>     outside
> 
> **Expected behavior**
> 
> No error is raised.
> 
> **Environment**
> 
>  - Vim version: 8.2 Included patches: 1-1253
>  - OS: Ubuntu 16.04.6 LTS
>  - Terminal: XTerm(356)
> 
> **Additional context**
> 
> The issue disappears inside a `:def` function:

In a compiled function variables are on the stack and it's relatively
easy to keep them local to a block scope.

At the script level variables are script variables.  We could make them
local to a block, which means we need to remember what variables were
declared and delete them when leaving the block.  While at the top
script level variables cannot be deleted.  Would that work intuitively?

	let var1 = 'at script level
	if {expr}
	   let var2 = 'block-local
	endif
	echo var1  # OK
	echo var2  # error!

I suppose that can work.

-- 
hundred-and-one symptoms of being an internet addict:
44. Your friends no longer send you e-mail...they just log on to your IRC
    channel.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by lacygoill** [comment]
> Would that work intuitively?

For me intuition is built upon repetition.  The more I repeat something, the more "intuitive" it gets.  But to easily repeat something, it must be consistent.  Right now, this looks inconsistent:
```vim
vim9script
def Func()
    let var1 = 'at script level'
    if 1
        let var2 = 'block-local'
    endif
    echo var1  # OK
    echo var2  # error!
enddef
defcompile
```
    Error detected while processing function <SNR>1_Func:
    line    6:
    E1001: variable not found: var2

```vim
vim9script
let var1 = 'at script level'
if 1
    let var2 = 'block-local'
endif
echo var1  # OK
echo var2  # error!
```
    at script level
    block-local

If we forget about the `:def` function, the 2 snippets are identical.  And yet, they give different results.  If it keeps working like this, I suspect I'll frequently make errors; therefore it will be harder to repeat this kind of code, and intuition will be harder to build.

If it can be made more consistent without having a negative impact on Vim9 script's performance, that would be great.  Otherwise, feel free to close the issue.

**Comment by lacygoill** [comment]
Now that I think about it more, I think you're right.  Those are completely different contexts:  function vs script.  Maybe it doesn't make sense to support this.  @bfrg:  what is your opinion?

**Comment by lacygoill** [comment]
Fixed by [8.2.1826](https://github.com/vim/vim/commit/9becdf2b98e56d5eb193f3413d706ea433269216).
